confluent_kafka/avro/serializer/MessageSerializer.py

```diff
@@ -0,0 +1,241 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+#
+# derived from https://github.com/verisign/python-confluent-schemaregistry.git
+#
+import io
+import logging
+import struct
+import sys
+import traceback
+import avro
+import avro.io
+
+from confluent_kafka.avro import ClientError
+from . import SerializerError
+
+MAGIC_BYTE = 0
+
+HAS_FAST = False
+try:
+    from fastavro.reader import read_data
+
+    HAS_FAST = True
+except:
+    pass
+
+
+class ContextStringIO(io.BytesIO):
+    """
+    Wrapper to allow use of StringIO via 'with' constructs.
+    """
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+        return False
+
+class MessageSerializer(object):
+    """
+    A helper class that can serialize and deserialize messages
+    that need to be encoded or decoded using the schema registry.
+
+    All encode_* methods return a buffer that can be sent to kafka.
+    All decode_* methods expect a buffer received from kafka.
+    """
+
+    def __init__(self, registry_client):
+        self.registry_client = registry_client
+        self.id_to_decoder_func = {}
+        self.id_to_writers = {}
+        self.log = logging.getLogger(__name__)
+
+    '''
+
+    '''
+
+    def encode_record_with_schema(self, topic, schema, record, is_key=False):
+        """
+        Given a parsed avro schema, encode a record for the given topic.  The
+        record is expected to be a dictionary.
+
+        The schema is registered with the subject of 'topic-value'
+        @:param topic : Topic name
+        @:param schema : Avro Schema
+        @:param record : A dictionary object
+        @:returns : Encoded record with schema ID as bytes
+        """
+        if not isinstance(record, dict):
+            self.log.error("record must be a dictionary")
+            raise SerializerError("record must be a dictionary")
+        subject_suffix = ('-key' if is_key else '-value')
+        # get the latest schema for the subject
+        subject = topic + subject_suffix
+        # register it
+        schema_id = self.registry_client.register(subject, schema)
+        if not schema_id:
+            message = "Unable to retrieve schema id for subject %s" % (subject)
+            self.log.error(message)
+            raise SerializerError(message)
+
+        # cache writer
+        self.id_to_writers[schema_id] = avro.io.DatumWriter(schema)
+
+        return self.encode_record_with_schema_id(schema_id, record)
+
+    # subject = topic + suffix
+    def encode_record_for_topic(self, topic, record, is_key=False):
```

Comment:
This seems like an odd use case -- it's basically assuming that the latest schema will be compatible with the data you pass in? Are you using this in practice?